modules/internal/DefaultRectangular.chpl

```diff
@@ -1007,6 +1115,10 @@ module DefaultRectangular {
     pragma "local field"
     var data : _ddata(eltType) = nil;
 
+    var externData: chpl_external_array;
+    var externArr: bool = false;
+    var _owned: bool = false;
```

Comment:
I'd maybe name the field "_borrowed" since setting it to true is the exception.

We have some compile-time analysis that relies on the current property that DefaultRectangular arrays never refer to the same memory (only array slices do). With this external array thing as implemented, that's no longer knowable at compile time. Would it possibly make sense for `externArr` to be a `param`? Or should we document the assumption the compiler makes - on the possible justification that arrays generally speaking refer to unique data?

Either way I'd put these `bool` fields after `no_init_data` for better packing.

Regarding the `externData` field, I don't believe this field needs to exist. Or at least I'd like to talk about that viewpoint. Doesn't DefaultRectangular already have what it needs to represent an external array?
 * External arrays are 1D (so a subset of DefaultRectangular arrays)
 * External arrays start at 0 (so a subset of DefaultRectangular arrays)
 * External arrays have a size / number of elements (but so do DefaultRectangular arrays)

It seems to me that the `chpl_free_func` is the only thing you'd need to add to be able to "save" a `chpl_external_array` in a DefaultRectangular and then re-construct the `chpl_external_array` later. And, I'm not sure we want to keep `chpl_free_func` if we are indeed moving towards building from the Fortran-C array interop API.

Reply:
> I'd maybe name the field "_borrowed" since setting it to true is the exception.

Can do!

> We have some compile-time analysis that relies on the current property that DefaultRectangular arrays never refer to the same memory (only array slices do). With this external array thing as implemented, that's no longer knowable at compile time. Would it possibly make sense for externArr to be a param? Or should we document the assumption the compiler makes - on the possible justification that arrays generally speaking refer to unique data?

externArr cannot currently be param due to the pattern where we create the domain and then assign its indices using `chpl_assignDomainWithGetSetIndices` - I chose to avoid reimplementing the support that function provides and instead turn externArr off and then on when it needs to be called.  It would have made some other things easier to have a different strategy for setting the indices and maybe I should pursue that still, but it didn't seem like the right tradeoff at the time.

> Either way I'd put these bool fields after no_init_data for better packing.

I can do that!

> Regarding the externData field, I don't believe this field needs to exist. Or at least I'd like to talk about that viewpoint. Doesn't DefaultRectangular already have what it needs to represent an external array?

I guess my main motivating use case was when an exported function takes and returns the same array (with some modifications, probably).  I'd like the wrapper to at least appear to the user as though it were identical - we can also achieve that by providing a comparison function that indicates two wrappers are the same when their memory pointer and size are the same, but there is a question of which version to clean up in that case (and should/could we protect the user from accidentally trying to free both)

Reply:
I'm sure there's a better strategy than toggling externArr when doing domain assignment. That sounds hacky anyway.

For the exported function taking and returning the same array - there are still 2 variables of type chpl_external_array (somewhere) - so you still have to solve the problem of "which one is responsible for freeing". I don't see why having this type literally in DefaultRectangular helps with that problem.

Reply:
Backing up like 9000 feet, I'd expect that "folding ExternArr into DefaultRectangular" would consist of:
 * adding a param field for is-it-an-extern-array
 * adding checks against that param field in various cases to make sure e.g. resizing the array isn't attempted.

I think it might be just as reasonable to let the array be resized (by freeing the extern array buffer) and become a regular DefaultRectangular array.

What other cases have divergent behavior for DR vs ExternalArray? 1-D or element type seem easy enough to handle at array creation (and don't even necessarily require any field at all).

Reply:
I think that's all the cases that differ.  That idea about resizing the array is interesting - I think it's a reasonable action item but probably beyond the scope of the current PR.

> I don't see why having this type literally in DefaultRectangular helps with that problem.

Fair, I suppose it will involve a copy anyways so it won't be literally the same instance.

> I'm sure there's a better strategy than toggling externArr when doing domain assignment. That sounds hacky anyway.

Yeah, it is hacky.  But arguably it is a little weird that creating the domain involves an assignment instead of creating it with the right indices to begin with.  I wasn't sure if that was done for optimization purposes, memory management "first-touch" help, or as yet another instance of the "assignment instead of init" pattern we dislike in our compiler so much, so it didn't seem reasonable to mess with it without understanding that fully.  My goal was to hopefully not mess up DefaultRectangular performance with this change and it seems like doing that in the same PR will make it harder to track down why performance would change if it does - with the PR in the current state, at least we know that any performance difference would come from messing up optimization analysis rather than a mix of that and intentionally changing behavior.  But since I'm going to run with the perf playground anyways, maybe having at least a night with the current set up and a night with that modification will be enough?

Reply:
Wait a minute... I think I missed something (I havn't tried to read all the code yet). Do you have a ExternDomain type represented here? I'd expect domains are just regular domains and DefaultRectangular(extern=true) verifies the domains are OK. I don't think you need an ExternDomain type at all. E.g. make dsiReallocate on extern=true arrays be an error.

Reply:
Nah, I merged the ExternDom implementation into DefaultRectangularDom.  It's a bit of a holdover from the original hierarchy but I think that knowledge needs to be in DefaultRectangularDom because of its dsiBuildArray and friends functions (i.e. I think dsiBuildArray needs to know when it is building an array to hold extern information otherwise it won't create things correctly - this will especially be true if we make externArr a param field)

Reply:
I don't think constructing an external array needs to call dsiBuildArray at all and in fact I think it'd be cleaner if it didn't do that.